Mathlib/Tactic/NormNum.lean

```diff
@@ -146,92 +147,140 @@ def eval (e : Expr) : MetaM Simp.Result := do
-    let prf ← derive ctx tgt
-    let newGoal ← applySimpResultToTarget mvarId tgt prf
-    let t ← inferType (mkMVar newGoal)
-    if t.isConstOf ``True then
-      newGoal.assign (mkConst ``True.intro)
-      return none
-    return some newGoal
-
-/-- Normnums a hypothesis. -/
-def normNumHyp (ctx : Simp.Context) (fvarId: FVarId) : TacticM Unit :=
-  liftMetaTactic1 fun mvarId => do
-    let hyp ← instantiateMVars (← fvarId.getDecl).type
-    let prf ← derive ctx hyp
-    let (some (_newHyp, newGoal)) ← applySimpResultToLocalDecl mvarId fvarId prf false
-      | throwError "Failed to apply norm_num to hyp."
-    return newGoal
-
-open Meta Elab Tactic in
-/-- Elaborator helper for norm num. -/
-def elabNormNum (args: Syntax) (loc : Syntax) : TacticM Unit := do
-  -- [todo] are there norm_num simp lemmas?
-  let simpTheorems ← ({} : SimpTheorems).addConst ``eq_self
-  let congrTheorems ← Meta.getSimpCongrTheorems
-  let eraseLocal := false
-  let kind := SimpKind.simp
-  let mut ctx : Simp.Context := {
-    config      := {}
-    simpTheorems := #[simpTheorems]
-    congrTheorems
-  }
-  let r ← Lean.Elab.Tactic.elabSimpArgs args (eraseLocal := eraseLocal) (kind := kind) ctx
-  ctx := r.ctx
-  let loc := expandOptLocation loc
-  withLocation loc
-    (atLocal := Meta.NormNum.normNumHyp ctx)
-    (atTarget := Meta.NormNum.normNumTarget ctx)
-    (failed := fun _ => throwError "norm_num failed")
+/--
+Constructs a proof that the original expression is true
+given a simp result which simplifies the target to `True`.
+-/
+def _root_.Lean.Meta.Simp.Result.ofTrue (r : Simp.Result) : MetaM (Option Expr) :=
+  if r.expr.isConstOf ``True then
+    some <$> match r.proof? with
+    | some proof => mkOfEqTrue proof
+    | none => pure (mkConst ``True.intro)
+  else
+    pure none
+
+/-- A simp plugin which calls `NormNum.eval`. -/
+def tryNormNum? (e : Expr) : SimpM (Option Simp.Step) :=
+  try return some (Simp.Step.done (← eval e)) catch _ => return none
+
+variable (ctx : Simp.Context) (useSimp := true) in
+mutual
+  /-- A discharger which calls `norm_num`. -/
+  partial def discharge (e : Expr) : SimpM (Option Expr) := do (← derive e).ofTrue
+
+  /-- The `Methods` to use when `useSimp` is enabled. -/
+  @[inline] partial def methodsSimp : Simp.Methods where
+    pre := (Simp.preDefault · discharge)
+    post e := do Simp.andThen (← Simp.postDefault e discharge) tryNormNum?
+    discharge? := discharge
+
+  /-- The `Methods` to use when `useSimp` is disabled. -/
+  @[inline] partial def methodsNoSimp : Simp.Methods where
+    post e := Simp.andThen (.visit { expr := e }) tryNormNum?
+    discharge? := discharge
+
+  /-- Traverses the given expression using simp and normalises any numbers it finds. -/
+  partial def derive (e : Expr) : MetaM Simp.Result :=
+    (·.1) <$> Simp.main e ctx (methods := if useSimp then methodsSimp else methodsNoSimp)
+end
+
+-- FIXME: had to inline a bunch of stuff from `simpGoal` here
```

Comment:
Just checking on your plan for these FIXMEs. Is it "merge to mathlib4 as is, and then eventually PR to core, but not yet"?

Reply:
Yes. I'm not in a hurry to PR this to core since there are a few other simp-based mathlib tactics and it seems reasonable to keep gathering data on how the core version should be structured to allow for `norm_num` and other tactics to use it, but I wouldn't object to it either.